### PR TITLE
Add Interlocked.{Compare}Exchange for UIntPtr

### LINKED ISSUE #45103: like #24694，Interlocked.CompareExchange/Exchange still missing for UIntPtr
## Background and Motivation

.net5 adds support for native int(nint and nuint), System.Threading.Interlocked should support them too

## Proposed API

```diff
namespace System.Threading
{
     public static class Interlocked {
+        public static extern UIntPtr CompareExchange(ref UIntPtr location1, UIntPtr value, UIntPtr comparand);
+        public static extern UIntPtr Exchange(ref UIntPtr location1, UIntPtr value);
     }
```

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/45103

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs [hunk 1]
```diff
@@ -83,6 +83,22 @@ public static IntPtr Exchange(ref IntPtr location1, IntPtr value)
...
+        /// <param name="location1">The variable to set to the specified value.</param>
+        /// <param name="value">The value to which the <paramref name="location1"/> parameter is set.</param>
+        /// <returns>The original value of <paramref name="location1"/>.</returns>
+        /// <exception cref="NullReferenceException">The address of location1 is a null pointer.</exception>
+        [CLSCompliant(false)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static UIntPtr Exchange(ref UIntPtr location1, UIntPtr value)
+        {
+#if TARGET_64BIT
+            return (UIntPtr)Interlocked.Exchange(ref Unsafe.As<UIntPtr, long>(ref location1), (long)value);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:97]: nit: Use `nuint` so this is an identity conversion and doesn't go through the user-defined cast operator.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:97]: I thought the plan was the C# compiler will soon start treating nint/IntPtr and nuint/UIntPtr as pure synonyms, at which point there shouldn't be such a difference.  Have I misunderstood?

I'm fine updating this if there's actually benefit, but for now I just stuck with consistency with the existing IntPtr overloads.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:97]: That's the plan yes. 

No strong preference on it actually happening if we'd rather wait for the compiler changes, just wanted to call it out as a current inefficiency.

**Review by tannergooding — APPROVED** [review]

**Comment by jkotas** [comment]
Windows arm64 tests fail due to known issue: https://github.com/dotnet/arcade/issues/9284

**Pull request merged**